lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,378 @@
+                      llvm::DenseMap<std::pair<StringRef, StringRef>,
+                                     Component::Info> *interfaceMap)
+      : options(opts), context(context), interfaceMap(interfaceMap) {}
+
+  using FIRRTLVisitor<GrandCentralVisitor>::visitDecl;
+
+  void visitDecl(FModuleOp op);
+  void visitDecl(FExtModuleOp op);
+  void visitDecl(RegOp op);
+  void visitDecl(WireOp op);
+  void visitDecl(NodeOp op);
+  void visitUnhandledOp(Operation *op) {}
+
+  ArrayAttr getAnnotations(Operation *op) {
+    return op->getAttrOfType<ArrayAttr>("annotations");
+  };
+
+private:
+  Options *options;
+  MLIRContext *context;
+  llvm::DenseMap<std::pair<StringRef, StringRef>, Component::Info>
+      *interfaceMap;
+
+  ArrayAttr handleRefLike(Operation *op, ArrayAttr annotations,
+                          FIRRTLType type);
+
+  void handleRef(Operation *op);
+
+  ArrayAttr handlePorts(Operation *op);
+};
+
+} // namespace
+
+void GrandCentralVisitor::visitDecl(FModuleOp op) {
+  op.setAllArgAttrs(handlePorts(op));
+
+  for (auto &stmt : op.getBodyBlock()->getOperations())
+    dispatchVisitor(&stmt);
+};
+
+void GrandCentralVisitor::visitDecl(FExtModuleOp op) {
+  op.setAllArgAttrs(handlePorts(op));
+};
+
+void GrandCentralVisitor::visitDecl(RegOp op) { handleRef(op); };
+
+void GrandCentralVisitor::visitDecl(WireOp op) { handleRef(op); };
+
+void GrandCentralVisitor::visitDecl(NodeOp op) { handleRef(op); };
+
+ArrayAttr GrandCentralVisitor::handleRefLike(mlir::Operation *op,
+                                             ArrayAttr annotations,
+                                             FIRRTLType type) {
+  if (!annotations)
+    return {};
+
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+
+  for (auto a : annotations) {
+    auto anno = a.cast<DictionaryAttr>();
+    auto clazz = anno.getAs<StringAttr>("class");
+
+    if (clazz == options->augmentedGroundType) {
+      auto defName = anno.getAs<StringAttr>("defName").getValue();
+      auto name = anno.getAs<StringAttr>("name").getValue();
+
+      // TODO: This is ignoring situations where the leaves of and interface are
+      // not ground types.  This enforces the requirement that this runs after
+      // LowerTypes.  However, this could eventually be relaxed.
+      if (!type.isGround()) {
+        // TODO: This diagnostic doesn't print as report_fatal_error causes a
+        // very quick exit. Fix this.
+        auto diag = op->emitOpError()
+                    << "cannot be added to interface '" << defName
+                    << "', component '" << name
+                    << "' because it is not a ground type. (Got type '" << type
+                    << "'.) This will be dropped from the interface. (Did you "
+                       "forget to run LowerTypes?)";
+        diag.attachNote()
+            << "The annotation indicating that this should be added was: '"
+            << a;
+        llvm::report_fatal_error(
```

Comment:
Please just report pass failure, instead of using `llvm::report_fatal_error` on invalid IR.

Reply:
This is tricky to untangle as this is inside a FIRRTLVisitor which isn't a pass and can't get at the protected `signalPassFailure`. I think the correct thing to do here is to change the return type of the visitors to not be `void` so that errors can then percolate back up.

Do you have a better way to architect this? (I was hoping for a single "get me out of here" call like `report_fatal_error`. 😄)

Reply:
Why not just set a bool (so the top level thing that created the visitor can know about the failure) and locally ignore the annotation?

Reply:
That works! I forget that mutable state is a thing. 🤔

Reply:
Changed to a mutable `bool failed = false` that, if set when the visitor gets examined, will cause a pass failure.